plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs

```diff
@@ -0,0 +1,96 @@
+module PlutusCoreInterpreter.CekMachine where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Evaluation.MachineException (MachineError (..), MachineException (..))
+import           Language.PlutusCore.Evaluation.Result           (EvaluationResult (..))
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import           Data.IntMap                                     (IntMap)
+import qualified Data.IntMap                                     as IntMap
+
+type Plain f = f TyName Name ()
+
+newtype Environment = Environment
+    { unEnvironment :: IntMap (Plain Value, Environment)
+    }
+
+data Frame
+    = FrameApplyFun Environment (Plain Value)    -- ^ @[V _]@
+    | FrameApplyArg Environment (Plain Term)     -- ^ @[_ N]@
+    | FrameTyInstArg (Type TyName ())            -- ^ @{_ A}@
+    | FrameUnwrap                                -- ^ @(unwrap _)@
+    | FrameWrap () (TyName ()) (Type TyName ())  -- ^ @(wrap α A _)@
+
+type Context = [Frame]
+
+extendEnvironment :: Name () -> Plain Value -> Environment -> Environment -> Environment
+extendEnvironment argName arg argEnv (Environment oldEnv) =
+    Environment $ IntMap.insert (unUnique $ nameUnique argName) (arg, argEnv) oldEnv
+
+lookupName :: Name () -> Environment -> Maybe (Plain Value, Environment)
+lookupName name (Environment env) = IntMap.lookup (unUnique $ nameUnique name) env
+
+computeCek :: Environment -> Context -> Plain Term -> EvaluationResult
+computeCek env con (TyInst _ fun ty)      = computeCek env (FrameTyInstArg ty : con) fun
+computeCek env con (Apply _ fun arg)      = computeCek env (FrameApplyArg env arg : con) fun
+computeCek env con (Wrap ann tyn ty term) = computeCek env (FrameWrap ann tyn ty : con) term
+computeCek env con (Unwrap _ term)        = computeCek env (FrameUnwrap : con) term
+computeCek env con tyAbs@TyAbs{}          = returnCek env con tyAbs
+computeCek env con lamAbs@LamAbs{}        = returnCek env con lamAbs
+computeCek env con constant@Constant{}    = returnCek env con constant
+computeCek _   _   Error{}                = EvaluationFailure
+computeCek env con var@(Var _ name)       = case lookupName name env of
+    Nothing           -> throw $ MachineException OpenTermEvaluatedMachineError var
```

Comment:
To elaborate on this: imagine someone using this machine to evaluate Plutus Core given to them by users. They do no want to catch exceptions, ideally they're crashing conditions indicating bugs. But the user might give them an open term, so they'd like to be robust to that and have it be an error they can handle normally.

Reply:
I think we should keep the CK machine as is. It's a reference interpreter and must be as simple as possible. Returning explicit errors would complicate it to a certain extent. But you're right, the CEK machine should handle this situation gracefully.
Then we also need to split exceptions into user and internal ones.

Reply:
Although... an open term won't type check, right?

Reply:
I'm not sure if an open term will typecheck currently, but we might want to allow it to. 

And maybe we won't always typecheck what the user gives us? I'd still be inclined to err on the side of returning errors due to malformed input too - they're "user errors" by the user of the function, and I'd still prefer to only throw if it's *our* error.

Re: user errors and internal ones, we can make the user error type an instance of `Exception` and then throw it in the CK machine and return it in the `CEK` machine :)

Reply:
We will always rename, though, because the CEK machine relies on a renamer pass. And we have there

```haskell
-- | A 'RenameError' is thrown when a free variable is encountered during
-- rewriting.
data RenameError a = UnboundVar (Name a)
                   | UnboundTyVar (TyName a)
                   deriving (Show, Eq, Generic, NFData)
```

So it's our error if there is a free variable and the renamer hasn't caught it.

Reply:
Okay, fair enough. Let's just make sure we don't expose an interface to a library user where they can get this wrong, then.

Reply:
What if someone wants to run a computation twice, but rename only once? Like when we put something on the blockchain (or is this not the case?).

Reply:
I thought about this a bit more and I think it doesn't make sense to return explicit errors in case an ill-typed terms was fed to the CK machine. Because
1) if you return an `Either`, it means the user must handle the `Left` case, i.e. handle the possibility of causing an internal error which is silly
2) the user is essentially forced to recheck that a term indeed was type checked before evaluated
3) the user can simply ignore the `Left` case and we won't know something goes wrong from time to time

There is nothing wrong about having invariants and punishing by throwing exceptions those who break them.

Reply:
I think your example of nodes maybe wanting to evaluate without typechecking is a good one. More generally, doing it that way gives people the choice. Given a CEK machine that returns an `Either`, we can then implement a version that doesn't return an either and instead throws, possibly requiring a typechecked term. And then users who don't want to do that naturally have to use the version that returns `Either`.

I just don't like locking the error handling strategy to "exception" unless they literally can only happen if they're a bug, which is not currently the case with the machine failures.

Reply:
> I just don't like locking the error handling strategy to "exception" unless they literally can only happen if they're a bug, which is not currently the case with the machine failures.

A renamer pass is a must, because otherwise you can get an `EvaluationSuccess` with a wrong result. Insisting on type checking terms is no harder than insisting on renaming them. I think making the type checker return a `WtTerm` turns machine failures into true bugs and resolves the issue.